File: Numerical_Methods_for_Physics_Garcia/edition2/Matlab/dsmceq.m
% dsmceq - Dilute gas simulation using DSMC algorithm
% This version illustrates the approach to equilibrium
clear all;  help dsmceq;   % Clear memory and print header

%* Initialize constants  (particle mass, diameter, etc.)
boltz = 1.3806e-23;    % Boltzmann's constant (J/K)
mass = 6.63e-26;       % Mass of argon atom (kg)
diam = 3.66e-10;       % Effective diameter of argon atom (m)
T = 273;               % Temperature (K)
density = 1.78;        % Density of argon at STP (kg/m^3)
L = 1e-6;              % System size is one micron
npart = input('Enter number of simulation particles: ');
eff_num = density/mass*L^3/npart;
fprintf('Each particle represents %g atoms\n',eff_num);

%* Assign random positions and velocities to particles
rand('state',0);       % Initialize random number generator
x = L*rand(npart,1);   % Assign random positions
v_init = sqrt(3*boltz*T/mass);    % Initial speed
v = zeros(npart,3);    % Only x-component is non-zero
v(:,1) = v_init * (1 - 2*floor(2*rand(npart,1)));

%* Plot the initial speed distribution
figure(1);  clf;
vmag = sqrt(v(:,1).^2 + v(:,2).^2 + v(:,3).^2);
vbin = 50:100:1050;    % Bins for histogram
hist(vmag,vbin);  title('Initial speed distribution');
xlabel('Speed (m/s)');  ylabel('Number');

%* Initialize variables used for evaluating collisions
ncell = 15;                     % Number of cells
tau = 0.2*(L/ncell)/v_init;     % Set timestep tau
vrmax = 3*v_init*ones(ncell,1); % Estimated max rel. speed
selxtra = zeros(ncell,1);       % Used by routine "colider"
coeff = 0.5*eff_num*pi*diam^2*tau/(L^3/ncell);
coltot = 0;                     % Count total collisions

%* Declare structure for lists used in sorting
sortData = struct('ncell',ncell,    ...
                  'npart',npart,    ...
                  'cell_n',zeros(ncell,1),  ...
                  'index',zeros(ncell,1),    ...
                  'Xref',zeros(npart,1));  

%* Loop for the desired number of time steps
nstep = input('Enter total number of time steps: ');
for istep = 1:nstep
	
  %* Move all the particles ballistically
  x(:) = x(:) + v(:,1)*tau; % Update x position of particle
  x = rem(x+L,L);           % Periodic boundary conditions

  %* Sort the particles into cells
  sortData = sorter(x,L,sortData);
  
  %* Evaluate collisions among the particles
  [v, vrmax, selxtra, col] = ...
                colider(v,vrmax,tau,selxtra,coeff,sortData);
  coltot = coltot + col;
  
  %* Periodically display the current progress
  if( rem(istep,10) < 1 )
    figure(2); clf;
    vmag = sqrt(v(:,1).^2 + v(:,2).^2 + v(:,3).^2);
    hist(vmag,vbin);
    title(sprintf('Done %g of %g steps; %g collisions',...
                                      istep,nstep,coltot));
    xlabel('Speed (m/s)');  ylabel('Number');
    drawnow;
  end
end

%* Plot the histogram of the final speed distribution
figure(2); clf;
vmag = sqrt(v(:,1).^2 + v(:,2).^2 + v(:,3).^2);
hist(vmag,vbin);  
title(sprintf('Final distrib., Time = %g sec.',nstep*tau));
xlabel('Speed (m/s)');  ylabel('Number');

File: Numerical_Methods_for_Physics_Garcia/edition2/python/dsmceq.py
import numpy as np
import matplotlib.pyplot as plt

# --- Constants ---
boltz = 1.3806e-23
mass  = 6.63e-26
diam  = 3.66e-10
T     = 273.0
density = 1.78
L     = 1e-6

#npart = int(input('Enter number of simulation particles: '))
npart = 10
eff_num = density / mass * L**3 / npart
print(f'Each particle represents {eff_num:.4g} atoms')

# --- Initial positions and velocities ---
rng = np.random.default_rng(0)
x = L * rng.random(npart)
v_init = np.sqrt(3 * boltz * T / mass)
v = np.zeros((npart, 3))
v[:, 0] = v_init * (1 - 2 * np.floor(2 * rng.random(npart)))

# --- Initial speed histogram ---
vmag = np.linalg.norm(v, axis=1)
vbin = np.arange(50, 1100, 100)
plt.figure(1)
plt.hist(vmag, bins=vbin)
plt.title('Initial speed distribution')
plt.xlabel('Speed (m/s)')
plt.ylabel('Number')
plt.tight_layout()
plt.show(block=False)

# --- Collision parameters ---
ncell   = 15
tau     = 0.2 * (L / ncell) / v_init
vrmax   = 3 * v_init * np.ones(ncell)
selxtra = np.zeros(ncell)
coeff   = 0.5 * eff_num * np.pi * diam**2 * tau / (L**3 / ncell)
coltot  = 0


def sorter(x, L, ncell, npart):
    """Sort particles into cells, return cell counts, start indices, and sorted refs."""
    cell_n = np.zeros(ncell, dtype=int)
    jc = np.floor(x / L * ncell).astype(int)
    jc = np.clip(jc, 0, ncell - 1)
    for i in range(npart):
        cell_n[jc[i]] += 1
    index = np.zeros(ncell, dtype=int)
    index[0] = 0
    for i in range(1, ncell):
        index[i] = index[i-1] + cell_n[i-1]
    # Build Xref: sorted particle indices by cell
    Xref = np.zeros(npart, dtype=int)
    count = index.copy()
    for i in range(npart):
        c = jc[i]
        Xref[count[c]] = i
        count[c] += 1
    return cell_n, index, Xref


def colider(v, vrmax, tau, selxtra, coeff, cell_n, index, Xref, ncell):
    """Perform DSMC collisions cell by cell."""
    col = 0
    for jcell in range(ncell):
        np_cell = cell_n[jcell]
        if np_cell < 2:
            continue
        # Number of candidate pairs to select
        select = coeff * np_cell * (np_cell - 1) * vrmax[jcell] + selxtra[jcell]
        nsel = int(select)
        selxtra[jcell] = select - nsel

        crmax = vrmax[jcell]
        start = index[jcell]

        for _ in range(nsel):
            # Pick two distinct random particles in this cell
            k  = int(np_cell * np.random.random())
            kk = int((np_cell - 1) * np.random.random())
            if kk >= k:
                kk += 1
            ip1 = Xref[start + k]
            ip2 = Xref[start + kk]

            # Relative speed
            cr = np.linalg.norm(v[ip1] - v[ip2])
            if cr > crmax:
                crmax = cr

            # Accept/reject collision
            if cr / vrmax[jcell] > np.random.random():
                col += 1
                # Update vrmax
                vrmax[jcell] = crmax
                # Center of mass velocity
                vcm = 0.5 * (v[ip1] + v[ip2])
                # Random unit vector
                cos_th = 1.0 - 2.0 * np.random.random()
                sin_th = np.sqrt(1.0 - cos_th**2)
                phi    = 2.0 * np.pi * np.random.random()
                vrel   = np.array([
                    cr * cos_th,
                    cr * sin_th * np.cos(phi),
                    cr * sin_th * np.sin(phi)
                ])
                v[ip1] = vcm + 0.5 * vrel
                v[ip2] = vcm - 0.5 * vrel

    return v, vrmax, selxtra, col


# --- Main time-stepping loop ---
#nstep = int(input('Enter total number of time steps: '))
nstep = 20
for istep in range(1, nstep + 1):
    # Move particles
    x += v[:, 0] * tau
    x  = np.mod(x + L, L)

    # Sort into cells
    cell_n, index, Xref = sorter(x, L, ncell, npart)

    # Collisions
    v, vrmax, selxtra, coltot_new = colider(
        v, vrmax, tau, selxtra, coeff, cell_n, index, Xref, ncell)
    coltot += coltot_new

    # Periodic display
    if istep % 10 == 0:
        vmag = np.linalg.norm(v, axis=1)
        plt.figure(2)
        plt.clf()
        plt.hist(vmag, bins=vbin)
        plt.title(f'Step {istep}/{nstep}; {coltot} collisions')
        plt.xlabel('Speed (m/s)')
        plt.ylabel('Number')
        plt.tight_layout()
        plt.pause(0.01)

# --- Final distribution ---
vmag = np.linalg.norm(v, axis=1)
plt.figure(2)
plt.clf()
plt.hist(vmag, bins=vbin)
plt.title(f'Final distribution, Time = {nstep*tau:.4g} s')
plt.xlabel('Speed (m/s)')
plt.ylabel('Number')
plt.tight_layout()
plt.savefig('dsmceq1.jpg')
